src/FilePositionMap.php

```diff
@@ -0,0 +1,146 @@
+    {
+        if ($node instanceof Token) {
+            $offset = $node->start;
+        } else {
+            $offset = $node->getStart();
+        }
+        return $this->getLineNumberForOffset($offset);
+    }
+
+    /**
+     * @param Node|Token $node
+     * Similar to getStartLine but includes the column
+     */
+    public function getStartLineCharacterPositionForOffset($node) : LineCharacterPosition
+    {
+        if ($node instanceof Token) {
+            $offset = $node->start;
+        } else {
+            $offset = $node->getStart();
+        }
+        return $this->getLineCharacterPositionForOffset($offset);
+    }
+
+    /** @param Node|Token $node */
+    public function getEndLine($node) : int
+    {
+        return $this->getLineNumberForOffset($node->getEndPosition());
+    }
+
+    /**
+     * @param Node|Token $node
+     * Similar to getStartLine but includes the column
+     */
+    public function getEndLineCharacterPositionForOffset($node) : LineCharacterPosition
+    {
+        return $this->getLineCharacterPositionForOffset($node);
+    }
+
+    /**
+     * @param Node|Token $node
+     * Similar to getStartLine but includes the column
+     */
+    public function getLineCharacterPositionForOffset(int $offset) : LineCharacterPosition
+    {
+        $line = $this->getLineNumberForOffset($offset);
+        $character = $this->getColumnForOffset($offset);
+        return new LineCharacterPosition($line, $character);
+    }
+
+    public function getLineNumberForOffset(int $offset) : int
+    {
+        if ($offset < 0) {
+            $offset = 0;
+        } elseif ($offset > $this->file_contents_length) {
+            $offset = $this->file_contents_length;
+        }
+        $current_offset = $this->current_offset;
+        if ($offset > $current_offset) {
+            $this->line_for_current_offset += \substr_count($this->file_contents, "\n", $current_offset, $offset - $current_offset);
+            $this->current_offset = $offset;
+        } elseif ($offset < $current_offset) {
+            $this->line_for_current_offset -= \substr_count($this->file_contents, "\n", $offset, $current_offset - $offset);
+            $this->current_offset = $offset;
+        }
+        return $this->line_for_current_offset;
+    }
+
+    /**
+     * @return int - gets the 1-based column offset
+     */
+    public function getColumnForOffset(int $offset) : int
+    {
+        $length = $this->file_contents_length;
+        if ($offset <= 1) {
+            return 1;
+        } elseif ($offset > $length) {
+            $offset = $length;
+        }
+        // Postcondition: offset >= 1, ($lastNewlinePos < $offset)
+        // If there was no previous newline, lastNewlinePos = 0
+
+        // Start strrpos check from the character before the current character,
+        // in case the current character is a newline.
+        $lastNewlinePos = \strrpos($this->file_contents, "\n", -$length + $offset - 1);
```

Comment:
When getting line and column, which is probably the most common case, we already know the position of the last newline. Wouldn't it be much faster to reuse that?

Reply:
I'd rather not focus on micro-optimizing right now, this is simpler and fairly fast.

I was using substr_count, which doesn't tell me **where** the last newline was. It just tells me how many newlines there were.

I'd expect the native string functions to be faster than doing a for loop in PHP.
I haven't yet tried repeatedly calling `stripos(contents, "\n")`. I imagine the other approach could be faster if the number of lines skipped over was small.

Reply:
That's fine